<a href="https://colab.research.google.com/github/simaphil/Image-Classification/blob/master/image_classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Import the dataset to a new file in tmp called weather

!mkdir '/tmp/weather'

!wget --no-check-certificate \
  https://md-datasets-cache-zipfiles-prod.s3.eu-west-1.amazonaws.com/4drtyfjtfy-1.zip -O \
  /tmp/weather/weather.zip

--2021-06-10 06:35:03--  https://md-datasets-cache-zipfiles-prod.s3.eu-west-1.amazonaws.com/4drtyfjtfy-1.zip
Resolving md-datasets-cache-zipfiles-prod.s3.eu-west-1.amazonaws.com (md-datasets-cache-zipfiles-prod.s3.eu-west-1.amazonaws.com)... 52.218.106.144
Connecting to md-datasets-cache-zipfiles-prod.s3.eu-west-1.amazonaws.com (md-datasets-cache-zipfiles-prod.s3.eu-west-1.amazonaws.com)|52.218.106.144|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 95592747 (91M) [application/octet-stream]
Saving to: ‘/tmp/weather/weather.zip’

/tmp/weather/weathe 100%[===================>]  91.16M  11.0MB/s    in 10s     

2021-06-10 06:35:14 (8.78 MB/s) - ‘/tmp/weather/weather.zip’ saved [95592747/95592747]



# Part 1: Splitting the data

In [ ]:
#Extract the data from the zip file
import os
import zipfile

local_zip = '/tmp/weather/weather.zip'
zip_ref = zipfile.ZipFile(local_zip, 'r')
zip_ref.extractall('/tmp/weather')
zip_ref.close()

local_zip = '/tmp/weather/dataset2.zip'
zip_ref = zipfile.ZipFile(local_zip, 'r')
zip_ref.extractall('/tmp/weather')
zip_ref.close()

#Get list of all files in directory
files=os.listdir('/tmp/weather/dataset2')

# The data has 1125 images:
#   300 cloudy,
#   215 rain,
#   253 shine,
#   357 sunrise

#Create folders for each category
!mkdir '/tmp/weather/dataset2/cloudy'
!mkdir '/tmp/weather/dataset2/rain'
!mkdir '/tmp/weather/dataset2/shine'
!mkdir '/tmp/weather/dataset2/sunrise'
#Move each image to the correct file:
for i in range(len(files)):
  p='/tmp/weather/dataset2/'+files[i] #Current image's path
  if(files[i][:4]=='rain'): #If the file name starts with rain move it to rain folder
    !mv {p} '/tmp/weather/dataset2/rain'
  else:
    if(files[i][:4]=='clou'):#If file name begins with clou move to cloudy folder
       !mv {p} '/tmp/weather/dataset2/cloudy'
    else:
      if(files[i][:4]=='shin'):#If file name begins with shin move to shine folder
          !mv {p} '/tmp/weather/dataset2/shine'
      else: #else move to sunrise folder
        !mv {p} '/tmp/weather/dataset2/sunrise'

In [ ]:
os.listdir('/tmp/weather/dataset2/cloudy')
#Show that list of files in folder is ordered randomly
#so that in future when I traverse it to split the data, it is done randomly

['cloudy238.jpg',
 'cloudy101.jpg',
 'cloudy194.jpg',
 'cloudy138.jpg',
 'cloudy84.jpg',
 'cloudy17.jpg',
 'cloudy127.jpg',
 'cloudy167.jpg',
 'cloudy203.jpg',
 'cloudy287.jpg',
 'cloudy285.jpg',
 'cloudy164.jpg',
 'cloudy208.jpg',
 'cloudy280.jpg',
 'cloudy297.jpg',
 'cloudy233.jpg',
 'cloudy292.jpg',
 'cloudy266.jpg',
 'cloudy291.jpg',
 'cloudy126.jpg',
 'cloudy171.jpg',
 'cloudy78.jpg',
 'cloudy247.jpg',
 'cloudy53.jpg',
 'cloudy49.jpg',
 'cloudy135.jpg',
 'cloudy12.jpg',
 'cloudy70.jpg',
 'cloudy119.jpg',
 'cloudy97.jpg',
 'cloudy213.jpg',
 'cloudy100.jpg',
 'cloudy281.jpg',
 'cloudy105.jpg',
 'cloudy185.jpg',
 'cloudy7.jpg',
 'cloudy212.jpg',
 'cloudy137.jpg',
 'cloudy146.jpg',
 'cloudy174.jpg',
 'cloudy24.jpg',
 'cloudy114.jpg',
 'cloudy44.jpg',
 'cloudy151.jpg',
 'cloudy67.jpg',
 'cloudy152.jpg',
 'cloudy56.jpg',
 'cloudy58.jpg',
 'cloudy244.jpg',
 'cloudy123.jpg',
 'cloudy112.jpg',
 'cloudy48.jpg',
 'cloudy28.jpg',
 'cloudy196.jpg',
 'cloudy242.jpg',
 'cloudy161.jpg',
 'cloudy2

In [ ]:
#Splitting the data to train and validation: I chose to do 80%/20% because it is common practice

!mkdir '/tmp/weather/dataset2/train' #Create a folder for train data
!mkdir '/tmp/weather/dataset2/validation' #Create a folder for validation data

#Create folders for each category inside train folder (this makes it supervised learning)
!mkdir '/tmp/weather/dataset2/train/cloudy'
!mkdir '/tmp/weather/dataset2/train/rain'
!mkdir '/tmp/weather/dataset2/train/shine'
!mkdir '/tmp/weather/dataset2/train/sunrise'

#Get the lists of files in each category
f_cloudy=os.listdir('/tmp/weather/dataset2/cloudy')
f_rain=os.listdir('/tmp/weather/dataset2/rain')
f_shine=os.listdir('/tmp/weather/dataset2/shine')
f_sunrise=os.listdir('/tmp/weather/dataset2/sunrise')

#Creating the training- 80% from each category
#(I showed above that the lists are ordered randomly)
for i in range(round(len(f_cloudy)*0.8)): #80% of cloudy images will be moved to train
#for i in range(round(len(f_cloudy)*0.7)): #Changing the slpit to 70%/30% to find best results
  p='/tmp/weather/dataset2/cloudy/'+f_cloudy[i]
  !mv {p} '/tmp/weather/dataset2/train/cloudy'

for i in range(round(len(f_rain)*0.8)): #80% of rain images will be moved to train
#for i in range(round(len(f_rain)*0.7)):  #Changing the slpit to 70%/30% to find best results
  p='/tmp/weather/dataset2/rain/'+f_rain[i]
  !mv {p} '/tmp/weather/dataset2/train/rain'

for i in range(round(len(f_shine)*0.8)): #80% of shine images will be moved to train
#for i in range(round(len(f_shine)*0.7)):  #Changing the slpit to 70%/30% to find best results
  p='/tmp/weather/dataset2/shine/'+f_shine[i]
  !mv {p} '/tmp/weather/dataset2/train/shine'

for i in range(round(len(f_sunrise)*0.8)): #80% of sunrise images will be moved to train
#for i in range(round(len(f_sunrise)*0.7)):  #Changing the slpit to 70%/30% to find best results
  p='/tmp/weather/dataset2/sunrise/'+f_sunrise[i]
  !mv {p} '/tmp/weather/dataset2/train/sunrise'

In [ ]:
#Creating the testing- 20% from each category

#Create folders for each category inside validation folder (this makes it supervised learning)
!mkdir '/tmp/weather/dataset2/validation/cloudy'
!mkdir '/tmp/weather/dataset2/validation/rain'
!mkdir '/tmp/weather/dataset2/validation/shine'
!mkdir '/tmp/weather/dataset2/validation/sunrise'

#Get the lists of files in each category
f_cloudy=os.listdir('/tmp/weather/dataset2/cloudy')
f_rain=os.listdir('/tmp/weather/dataset2/rain')
f_shine=os.listdir('/tmp/weather/dataset2/shine')
f_sunrise=os.listdir('/tmp/weather/dataset2/sunrise')

for i in f_cloudy: #The images left in cloudy folder (20%) will be moved to validation folder
  p='/tmp/weather/dataset2/cloudy/'+i
  !mv {p} '/tmp/weather/dataset2/validation/cloudy'

for i in f_rain:
  p='/tmp/weather/dataset2/rain/'+i #The images left in rain folder (20%) will be moved to validation folder
  !mv {p} '/tmp/weather/dataset2/validation/rain'

for i in f_shine:
  p='/tmp/weather/dataset2/shine/'+i #The images left in shine folder (20%) will be moved to validation folder
  !mv {p} '/tmp/weather/dataset2/validation/shine'

for i in f_sunrise:
  p='/tmp/weather/dataset2/sunrise/'+i #The images left in sunrise folder (20%) will be moved to validation folder
  !mv {p} '/tmp/weather/dataset2/validation/sunrise'

In [ ]:
train_dir ='/tmp/weather/dataset2/train'
validation_dir = '/tmp/weather/dataset2/validation'

In [ ]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator
train_datagen = ImageDataGenerator(
    rescale=1./255)

val_datagen = ImageDataGenerator(rescale=1./255)

# Flow training images in batches of 32 using train_datagen generator
train_generator = train_datagen.flow_from_directory(
        train_dir,  # This is the source directory for training images
        target_size=(150, 150),  # All images will be resized to 150x150
        batch_size=35,
        # Since we use binary_crossentropy loss, we need binary labels
        class_mode='categorical')

# Flow validation images in batches of 32 using val_datagen generator
validation_generator = val_datagen.flow_from_directory(
        validation_dir,
        target_size=(150, 150),
        batch_size=35,
        class_mode='categorical')

Found 809 images belonging to 4 classes.
Found 316 images belonging to 4 classes.


When I used 80%/20% splitting the results were:

Found 900 images belonging to 4 classes.

Found 225 images belonging to 4 classes.

When I used 70%30%:

Found 809 images belonging to 4 classes.

Found 316 images belonging to 4 classes.

# Part 2: CNN

In [ ]:
from tensorflow.keras import layers
from tensorflow.keras import Model
from tensorflow.keras.optimizers import RMSprop

#I chose to use the CNN from targil 5 with modifications,
#the main one is using softmax in output layer instead of sigmoid

# Our input feature map is 150x150x3: 150x150 for the image pixels, and 3 for
# the three color channels: R, G, and B
from tensorflow import keras

model = keras.Sequential(
    [
        layers.Input(shape=(150, 150, 3)),
        # First convolution extracts 16 filters that are 3x3
        # Convolution is followed by max-pooling layer with a 2x2 window
        layers.Conv2D(16, 3, activation='relu'),
        layers.MaxPooling2D(2),
        # Second convolution extracts 32 filters that are 3x3
        # Convolution is followed by max-pooling layer with a 2x2 window
        layers.Conv2D(32, 3, activation='relu'),
        layers.MaxPooling2D(2),
        # Third convolution extracts 64 filters that are 3x3
        # Convolution is followed by max-pooling layer with a 2x2 window
        layers.Conv2D(64, 3, activation='relu'),
        layers.MaxPooling2D(2),
        # Flatten feature map to a 1-dim tensor so we can add fully connected layers
        layers.Flatten(),
        layers.Dropout(0.5),
        layers.Dense(256, activation='relu'),
        layers.Dense(4, activation='softmax'), #4 categories
    ]
)

In [ ]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_3 (Conv2D)            (None, 146, 146, 16)      1216      
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 73, 73, 16)        0         
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 71, 71, 32)        4640      
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 35, 35, 32)        0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 33, 33, 64)        18496     
_________________________________________________________________
max_pooling2d_5 (MaxPooling2 (None, 16, 16, 64)        0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 16384)            

In [ ]:
from tensorflow.keras.optimizers import RMSprop,Adam, SGD

model.compile(loss='binary_crossentropy',
              optimizer=SGD(lr=0.001,momentum=0.9),
              metrics=['acc'])

In [ ]:
history = model.fit_generator(train_generator, steps_per_epoch=len(train_generator),
		validation_data=validation_generator, validation_steps=len(validation_generator), epochs=24, verbose=2)
	# evaluate model

In [ ]:
_, acc = model.evaluate_generator(validation_generator, steps=len(validation_generator), verbose=0)
print('> %.3f' % (acc * 100.0))

Results:

First run: Batch size=32, number of epochs=15, dropout=0.5, optimizer=RMS with lr=0.001, accuracy= **91.15%**

Second run: Batch size=20, number of epochs=10, dropout=0.5, optimizer=RMS with lr=0.001, accuracy= **94.69%**

Third run: Batch size=20, number of epochs=10, dropout=0.5, optimizer=RMS with lr=0.01, accuracy= **63.274%**

Fourth run: Batch size=20, number of epochs=10, dropout=0.25, optimizer=RMS with lr=0.001, accuracy= **92.444%**

Fifth run: Batch size=40, number of epochs=20, dropout=0.5, optimizer=RMS with lr=0.001, data splitting 70%/30%, accuracy= **93.038%**

Sixth run: Batch size=35, number of epochs=20, dropout=0.5, optimizer=Adam with lr=0.001 and beta_1=0.9, data splitting 70%/30%, accuracy= **94.937%**

Seventh run: Batch size=35, number of epochs=25, dropout=0.5, optimizer=SGD with lr=0.001 and momentum=0.9, data splitting 70%/30%, accuracy= **94.62%**


# Part 3: Adding data augmentation

In [ ]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator
# Adding rescale, rotation_range, width_shift_range, height_shift_range,
# shear_range, zoom_range, and horizontal flip to our ImageDataGenerator
train_datagen = ImageDataGenerator(
    rescale=1./255,
    rotation_range=38,
    width_shift_range=0.4, #previous value: 0.2
    height_shift_range=0.2, #previous value: 0.2
    shear_range=0.35, #previous value: 0.2
    zoom_range=0.27, #previous value: 0.2
    horizontal_flip=True,)

val_datagen = ImageDataGenerator(rescale=1./255)

# Flow training images in batches of 32 using train_datagen generator
train_generator = train_datagen.flow_from_directory(
        train_dir,  # This is the source directory for training images
        target_size=(150, 150),  # All images will be resized to 150x150
        batch_size=35,
        # Since we use binary_crossentropy loss, we need binary labels
        class_mode='categorical')

# Flow validation images in batches of 32 using val_datagen generator
validation_generator = val_datagen.flow_from_directory(
        validation_dir,
        target_size=(150, 150),
        batch_size=35,
        class_mode='categorical')

Found 900 images belonging to 4 classes.
Found 225 images belonging to 4 classes.


In [ ]:
from tensorflow.keras.optimizers import RMSprop,Adam,SGD

model.compile(loss='binary_crossentropy',
              optimizer=SGD(lr=0.0018,momentum=0.9),
              metrics=['acc'])

In [ ]:
history = model.fit_generator(train_generator, steps_per_epoch=len(train_generator),
		validation_data=validation_generator, validation_steps=len(validation_generator), epochs=24, verbose=2)
	# evaluate model

In [ ]:
_, acc = model.evaluate_generator(validation_generator, steps=len(validation_generator), verbose=0)
print('> %.3f' % (acc * 100.0))

Results:

First run: Batch size=20, number of epochs=15, dropout=0.5, optimizer=RMS with lr=0.001, accuracy= **92.889%**

Second run: Batch size=32, number of epochs=10, dropout=0.5, optimizer=RMS with lr=0.001, changes made to values in the augmentation definition, accuracy= **93.778%**

Third run: same as second but with batch size=20, accuracy=**92.889%**

Fourth run: same as third but with batch size=32, optimizer=Adam, accuracy=**93.778%**

Fifth run: Batch size=35, number of epochs=25, dropout=0.5, optimizer=Adam, data splitting 70%/30%, accuracy= **92.722%**

Fifth run: Batch size=35, number of epochs=25, dropout=0.5, optimizer=SGD, data splitting 70%/30%, accuracy= **95.886%**

Sixth run: Batch size=35, number of epochs=24, dropout=0.5, optimizer=SGD, data splitting 70%/30%, I changed the values of the augmentation and SGD, accuracy= **96.519%**

Conclusion: Data augmentation improves results.


# Part 4: Transfer learning

In [ ]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator
#Using data augmentation
train_datagen = ImageDataGenerator(
    rescale=1./255,
     rotation_range=43,
    width_shift_range=0.24, #previous value: 0.2
    height_shift_range=0.125, #previous value: 0.2
    shear_range=0.55, #previous value: 0.2
    zoom_range=0.15, #previous value: 0.2
    horizontal_flip=True,)

val_datagen = ImageDataGenerator(rescale=1./255)

# Flow training images in batches of 32 using train_datagen generator
train_generator = train_datagen.flow_from_directory(
        train_dir,  # This is the source directory for training images
        target_size=(224,224),  # All images will be resized to 224x224 for VGG
        batch_size=12,
        # Since we use binary_crossentropy loss, we need binary labels
        class_mode='categorical')

# Flow validation images in batches of 32 using val_datagen generator
validation_generator = val_datagen.flow_from_directory(
        validation_dir,
        target_size=(224,224),
        batch_size=9,
        class_mode='categorical')

Found 900 images belonging to 4 classes.
Found 225 images belonging to 4 classes.


In [ ]:
#Transfer learning with VGG16
from tensorflow.keras.applications.vgg16 import VGG16
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Dense, Activation, Flatten,Dropout
from tensorflow.keras.optimizers import SGD,Adam,RMSprop
model = VGG16(include_top=False, input_shape=(224, 224, 3))
	# mark loaded layers as not trainable
for layer in model.layers:
		layer.trainable = False
	# add new classifier layers
dropout1= Dropout(0.15)(model.layers[-1].output) #adding dropout
flat1 = Flatten()(dropout1)
class1 = Dense(512, activation='relu', kernel_initializer='he_uniform')(flat1)
output = Dense(4, activation='softmax')(class1) #softmax activation in output layer
# define new model
model = Model(inputs=model.inputs, outputs=output)
# compile model
opt =Adam( lr=0.0018,beta_1=0.9)
model.compile(optimizer=opt, loss='binary_crossentropy', metrics=['accuracy'])

In [ ]:
history = model.fit_generator(train_generator, steps_per_epoch=len(train_generator),
		validation_data=validation_generator, validation_steps=len(validation_generator), epochs=40, verbose=2)
	# evaluate model
_, acc = model.evaluate_generator(validation_generator, steps=len(validation_generator), verbose=0)
print('> %.3f' % (acc * 100.0))


Results:

First run: batch size=20, number of epochs=10, dropout=0, optimizer=SGD with lr=0.001 and momentum=0.9, with augmentation, accuracy= **91.111%**

Second run: batch size=20, number of epochs=10, dropout=0.25, optimizer=SGD with lr=0.001 and momentum=0.9, with augmentation, accuracy= **90.222%**

Third run: same as second but no dropout and no augmentation, accuracy= **88.444%**

Fourth run: batch size=20, number of epochs=15, dropout=0, optimizer=SGD with lr=0.001 and momentum=0.9, with augmentation, accuracy=**92.444%**

Fifth run: batch size=20, number of epochs=15, dropout=0, optimizer=Adam with lr=0.001 and beta_1=0.9, with augmentation, accuracy= **93.778%**

Sixth run: same as fifth with batch size=32, epochs=20, accuracy= **95.111%**

Seventh run: same as sixth, 22 epochs,accuracy= **95.556%**

Eighth run: same as sixth but optimizer=RMSprop, accuracy=**92.444%**

Ninth run: batch size=32, number of epochs=22, dropout=0, optimizer=Adam with lr=0.001 and beta_1=0.9, with augmentation, with data splitting 70%/30% instead of 20%/80%, accuracy= **91.139%**

Tenth run: batch size=25, number of epochs=20, dropout=0.25, optimizer=Adam with lr=0.0018 and beta_1=0.92, with augmentation, with data splitting 80%/20%, accuracy= **95.556%**

Eleventh run: batch size=11, number of epochs=35, dropout=0.15, optimizer=Adam with lr=0.0018 and beta_1=0.92, with augmentation, with data splitting 80%/20%, accuracy= **96%**